compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -129,50 +129,69 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   *  }
+   *
+   *  def x(): Int = if (x$lzy.initialized()) x$lzy.value() else x$lzycompute()
+   *  ```
+   *
+   *  TODO: Implement Unit-typed lazy val optimization described above
+   */
   def transformLocalDef(x: ValOrDefDef)(implicit ctx: Context): Thicket = {
-      val valueInitter = x.rhs
-      val xname = x.name.asTermName
-      val holderName = LazyLocalName.fresh(xname)
-      val initName = LazyLocalInitName.fresh(xname)
-      val tpe = x.tpe.widen.resultType.widen
-
-      val holderType =
-        if (tpe isRef defn.IntClass) "LazyInt"
-        else if (tpe isRef defn.LongClass) "LazyLong"
-        else if (tpe isRef defn.BooleanClass) "LazyBoolean"
-        else if (tpe isRef defn.FloatClass) "LazyFloat"
-        else if (tpe isRef defn.DoubleClass) "LazyDouble"
-        else if (tpe isRef defn.ByteClass) "LazyByte"
-        else if (tpe isRef defn.CharClass) "LazyChar"
-        else if (tpe isRef defn.ShortClass) "LazyShort"
-        else "LazyRef"
-
-
-      val holderImpl = ctx.requiredClass("dotty.runtime." + holderType)
-
-      val holderSymbol = ctx.newSymbol(x.symbol.owner, holderName, containerFlags, holderImpl.typeRef, coord = x.pos)
-      val initSymbol = ctx.newSymbol(x.symbol.owner, initName, initFlags, MethodType(Nil, tpe), coord = x.pos)
-      val result = ref(holderSymbol).select(lazyNme.value).withPos(x.pos)
-      val flag = ref(holderSymbol).select(lazyNme.initialized)
-      val initer = valueInitter.changeOwnerAfter(x.symbol, initSymbol, this)
-      val initBody =
-        adaptToType(
-          ref(holderSymbol).select(defn.Object_synchronized).appliedTo(
-            adaptToType(mkNonThreadSafeDef(result, flag, initer, nullables = Nil), defn.ObjectType)),
-          tpe)
-      val initTree = DefDef(initSymbol, initBody)
-      val holderTree = ValDef(holderSymbol, New(holderImpl.typeRef, List()))
-      val methodBody = tpd.If(flag.ensureApplied,
-        result.ensureApplied,
-        ref(initSymbol).ensureApplied).ensureConforms(tpe)
-
-      val methodTree = DefDef(x.symbol.asTerm, methodBody)
-      ctx.debuglog(s"found a lazy val ${x.show},\nrewrote with ${holderTree.show}")
-      Thicket(holderTree, initTree, methodTree)
+    val xname = x.name.asTermName
+    val holderName = LazyLocalName.fresh(xname)
+    val initName = LazyLocalInitName.fresh(xname)
+    val tpe = x.tpe.widen.resultType.widen
+
+    val holderType =
+      if (tpe isRef defn.IntClass) "LazyInt"
+      else if (tpe isRef defn.LongClass) "LazyLong"
+      else if (tpe isRef defn.BooleanClass) "LazyBoolean"
+      else if (tpe isRef defn.FloatClass) "LazyFloat"
+      else if (tpe isRef defn.DoubleClass) "LazyDouble"
+      else if (tpe isRef defn.ByteClass) "LazyByte"
+      else if (tpe isRef defn.CharClass) "LazyChar"
+      else if (tpe isRef defn.ShortClass) "LazyShort"
+      else "LazyRef"
+
+    // val x$lzy = new scala.runtime.LazyInt()
+    val holderImpl = ctx.requiredClass("scala.runtime." + holderType)
```

Comment:
Instead of computing this over and over, I suggest moving it to Definitions.scala